proxy/envoy/config.go

```diff
@@ -126,9 +124,75 @@ func Generate(instances []*model.ServiceInstance, services []*model.Service, rul
 // buildListeners uses iptables port redirect to route traffic either into the
 // pod or outside the pod to service clusters based on the traffic metadata.
 func buildListeners(instances []*model.ServiceInstance,
-	services []*model.Service,
+	services []*model.Service, rules []*config.RouteRule,
+	upstreams []*config.UpstreamCluster,
 	mesh *MeshConfig) ([]Listener, []Cluster) {
-	clusters := buildClusters(services)
+
+	// services holds the default k8s services
+	// upstreams holds the different versions of the services specified by the user
+	// upstreams takes priority over services. We create one canonical set of
+	// service objects and pass it to the buildClusters function
+	// Create maps of all the arrays for easier cross referencing
+	serviceMapTmp := make(map[string]*model.Service, len(services))
+	for _, svc := range services {
+		serviceMapTmp[svc.Hostname] = svc
+	}
+
+	// Rules indexed by service name/hostname/destination
+	// There can be more than one rule per destination
```

Comment:
Please guarantee unique ordering everywhere!
It's not clear to me that the config is idempotent on relisting.